### PR TITLE
WIP: Upgrade cas es (Don't merge)

### PR DESCRIPTION
**Pull request by e-orz**

### CODE REVIEW

File: server/build.sbt [hunk 1]
```diff
@@ -45,6 +45,7 @@ initialize := {
     sys.error("Java 8 or higher is required for CM-Well!")
 }
 //resolvers in Global += "CM-WELL public" at "http://builder.clearforest.com:8081/nexus/content/groups/public"
+//resolvers in ThisBuild += "Local Maven" at Path.userHome.asFile.toURI.toURL + ".m2/repository"
```
Comment (bryaakov) [inline server/build.sbt]: remove those two (commented out) lines

File: server/cmwell-ws/app/cmwell/ws/Streams.scala [hunk 2]
```diff
@@ -169,6 +180,7 @@ class Streams @Inject()(crudServiceFS: CRUDServiceFS) extends LazyLogging {
     }
   }
 
+/*
```
Comment (bryaakov) [inline server/cmwell-ws/app/cmwell/ws/Streams.scala]: remove commented out code

File: server/cmwell-ws/app/cmwell/ws/Streams.scala [hunk 3]
```diff
@@ -310,7 +307,7 @@ class Streams @Inject()(crudServiceFS: CRUDServiceFS) extends LazyLogging {
   )(implicit ec: ExecutionContext): Future[(Source[IterationResults, NotUsed], Long)] = {
 
     val firstHitsTuple = crudServiceFS
-      .startScroll(
+      .startScrollEliNew(
```
Comment (bryaakov) [inline server/cmwell-ws/app/cmwell/ws/Streams.scala]: remove EliNew

File: server/cmwell-ws/app/controllers/Application.scala [hunk 4]
```diff
@@ -707,36 +706,41 @@ callback=< [URL] >
               }
 
               val fmFut = extractFieldsMask(request, typesCache, cmwellRDFHelper, timeContext)
+              val debugInfoParam = request.queryString.keySet("debug-info")
               crudServiceFS
-                .startScroll(
+                .thinSearch(
```
Comment (bryaakov) [inline server/cmwell-ws/app/controllers/Application.scala]: we should provide length=1 in this case, shouldn't we? _//cc @lielShraga_

File: server/cmwell-ws/app/ld/cmw/TimeBasedAccumulatedNamespacesCache.scala [hunk 5]
```diff
@@ -180,7 +180,7 @@ class TimeBasedAccumulatedNsCache private (private[this] var mainCache: Map[NsID
       Future.successful(sb.append('}').result())
     } else (actor ? GetStatus).mapTo[String]
 
-  @inline override def init(time: Long): Unit = actor ! UpdateRequest(time)
+  @inline override def init(time: Long): Unit = {} //actor ! UpdateRequest(time)
```
Comment (bryaakov) [inline server/cmwell-ws/app/ld/cmw/TimeBasedAccumulatedNamespacesCache.scala]: let's try to revive this `init`

File: server/cmwell-ws/app/logic/CRUDServiceFS.scala [hunk 6]
```diff
@@ -800,7 +800,7 @@ class CRUDServiceFS @Inject()(implicit ec: ExecutionContext, sys: ActorSystem) e
 
   private def toFieldValues(ss: Set[String]): Set[FieldValue] = ss.map(FString.apply)
 
-  def startScroll(pathFilter: Option[PathFilter] = None,
+  def startScrollEliNew(pathFilter: Option[PathFilter] = None,
```
Comment (bryaakov) [inline server/cmwell-ws/app/logic/CRUDServiceFS.scala]: please `sed s/EliNew//g` all over the place... :)

File: server/cmwell-ws/app/logic/CRUDServiceFS.scala [hunk 7]
```diff
@@ -1054,7 +1054,7 @@ class CRUDServiceFS @Inject()(implicit ec: ExecutionContext, sys: ActorSystem) e
                   val prevInfoton = infpot.getOrElse(
                     throw new RuntimeException(s"Previous infoton for path $path was not found under uuid ${pv.uuid}")
                   )
-                  ftsService.purgeByUuidsFromAllIndexes(Vector(pv.uuid)).flatMap { _ =>
+                  ftsService.purgeByUuidsFromAllIndexes(Vector(pv.uuid), partition = "blahblah").flatMap { _ =>
```
Comment (bryaakov) [inline server/cmwell-ws/app/logic/CRUDServiceFS.scala]: ⚠️ remove blah blah

File: server/cmwell-ws/src/pack/application.conf [hunk 8]
```diff
@@ -105,7 +105,7 @@ cmwell {
       }
     }
 
-    switch-over-to-zstore.file-size = 512K
+    switch-over-to-zstore.file-size = 256K
```
Comment (bryaakov) [inline server/cmwell-ws/src/pack/application.conf]: add comment: We have another one like that somewhere else

File: server/cmwell-bg/build.sbt [hunk 9]
```diff
@@ -25,14 +25,22 @@ libraryDependencies ++= {
...
-    dm("uk.org.lidalia","sysout-over-slf4j")
-  )
+    dm("uk.org.lidalia","sysout-over-slf4j"),
+    dm("io.netty", "netty-transport")
+  ).map{_.exclude("io.netty", "netty")}
 }
 
+libraryDependencies ++= Seq(
+  "net.lingala.zip4j" % "zip4j" % "1.3.2" % Test,
+  "pl.allergro.tech" % "embedded-elasticsearch" % "2.4.3" % Test from "https://github.com/israel/embedded-elasticsearch/releases/download/2.4.3/embedded-elasticsearch-2.4.3.jar"
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-bg/build.sbt]: done

File: server/cmwell-bg/build.sbt [hunk 10]
```diff
@@ -25,14 +25,22 @@ libraryDependencies ++= {
...
       .exclude("org.slf4j", "slf4j-log4j12")
       .exclude("log4j", "log4j"),
-    dm("uk.org.lidalia","sysout-over-slf4j")
-  )
+    dm("uk.org.lidalia","sysout-over-slf4j"),
+    dm("io.netty", "netty-transport")
+  ).map{_.exclude("io.netty", "netty")}
 }
 
+libraryDependencies ++= Seq(
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-bg/build.sbt]: done

File: server/cmwell-bg/src/test/scala/cmwell/bg/test/BGSpecs.scala [hunk 11]
```diff
@@ -14,16 +14,57 @@
...
+      println("Starting EmbeddedElasticSearch before all test suites")
+      EmbeddedES.embeddedElastic.start()
+    } else {
+      EmbeddedES.embeddedElastic.recreateIndices()
+    }
+    super.beforeAll()
+  }
+
+  override def afterAll(): Unit = {
+    if(nestedSuites.nonEmpty) {
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-bg/src/test/scala/cmwell/bg/test/BGSpecs.scala]: The file is already removed.

File: server/cmwell-common/src/test/scala/cmwell/common/InfotonSerializationSpecification.scala [hunk 12]
```diff
@@ -112,13 +114,13 @@ object InfotonSerializationSpecification extends Properties("Infoton") {
...
   }
 
-  property("encodeAndDecode") = forAll(infotons) {
-    i: Infoton => {
-      val e = JS.encodeInfoton(i)
-      val d = JS.decodeInfoton(e)
-      iCmp(i, d)
-    }
-  }
+//  property("encodeAndDecode") = forAll(infotons) {
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-common/src/test/scala/cmwell/common/InfotonSerializationSpecification.scala]: doen

File: server/cmwell-cons/app/scripts/templates/cassandra.yaml [hunk 13]
```diff
@@ -17,6 +17,9 @@ data_file_directories:
     - {{root_dir}}/data/{{dir}}/data
 
 commitlog_directory: {{root_dir}}/data/{{ccl_dir}}/commitlog
+hints_directory: {{root_dir}}/data/{{ccl_dir}}/hints
+cdc_enabled: false
+cdc_raw_directory: {{root_dir}}/data/{{ccl_dir}}/cdc_raw
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-cons/app/scripts/templates/cassandra.yaml]: To be fixed with another commit

File: server/cmwell-ws/app/logic/CRUDServiceFS.scala [hunk 14]
```diff
@@ -783,7 +781,7 @@ class CRUDServiceFS @Inject()(implicit ec: ExecutionContext, sys: ActorSystem) e
     }
 
     ftsService
-      .getLastIndexTimeFor(dc, withHistory = withHistory, fieldFilters = fieldFilters)
+      .getLastIndexTimeFor(dc, withHistory = withHistory, fieldFilters = fieldFilters, partition = "blahblah")
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-ws/app/logic/CRUDServiceFS.scala]: done

File: server/cmwell-ws/app/logic/CRUDServiceFS.scala [hunk 15]
```diff
@@ -920,17 +918,18 @@ class CRUDServiceFS @Inject()(implicit ec: ExecutionContext, sys: ActorSystem) e
...
     irwService.purgeHistorical(infoton, isOnlyVersion = false, QUORUM).flatMap { _ =>
-      ftsService.purge(infoton.uuid).map(_ => Unit)
+//        ftsService.purge(infoton.uuid).map(_ => Unit)
+      ???
     }
   }
 
   def purgeUuidFromIndex(uuid: String, index: String): Future[Unit] = {
-    ftsService.purgeByUuidsAndIndexes(Vector(uuid -> index)).map(_ => ()) //TODO also purge from ftsServiceNew
+    ftsService.purgeByUuidsAndIndexes(Vector(uuid -> index), partition = "blahblah").map(_ => ()) //TODO also purge from ftsServiceNew
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-ws/app/logic/CRUDServiceFS.scala]: done

File: server/cmwell-ws/app/logic/CRUDServiceFS.scala [hunk 16]
```diff
@@ -958,7 +957,7 @@ class CRUDServiceFS @Inject()(implicit ec: ExecutionContext, sys: ActorSystem) e
 
             val purgeJustByUuids = {
               if (justUuids.nonEmpty)
-                ftsService.purgeByUuidsFromAllIndexes(justUuids.toVector)
+                ftsService.purgeByUuidsFromAllIndexes(justUuids.toVector, partition = "blahblah")
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-ws/app/logic/CRUDServiceFS.scala]: done

File: server/cmwell-ws/app/logic/CRUDServiceFS.scala [hunk 17]
```diff
@@ -970,7 +969,7 @@ class CRUDServiceFS @Inject()(implicit ec: ExecutionContext, sys: ActorSystem) e
                 )
               } else {
                 if (uuidsWithIndexes.nonEmpty)
-                  ftsService.purgeByUuidsAndIndexes(uuidsWithIndexes.toVector)
+                  ftsService.purgeByUuidsAndIndexes(uuidsWithIndexes.toVector, partition = "blahblah")
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-ws/app/logic/CRUDServiceFS.scala]: done

File: server/cmwell-ws/app/logic/CRUDServiceFS.scala [hunk 18]
```diff
@@ -1079,7 +1080,7 @@ class CRUDServiceFS @Inject()(implicit ec: ExecutionContext, sys: ActorSystem) e
       val uuids = casHistory.map(_._2)
 
       retry(3, 1.seconds) {
-        val purgeEsByUuids = ftsService.purgeByUuidsFromAllIndexes(uuids)
+        val purgeEsByUuids = ftsService.purgeByUuidsFromAllIndexes(uuids, partition = "blahblah")
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-ws/app/logic/CRUDServiceFS.scala]: done

File: server/build.sbt (hunk 1 above)
[the remark this answers is not in the document]
Comment (e-orz) [reply server/build.sbt]: done

File: server/build.sbt [hunk 19]
```diff
@@ -57,56 +58,60 @@ dependenciesManager in Global := {
...
   case ("com.spatial4j","spatial4j")                               => "com.spatial4j" % "spatial4j" % "0.5"
   case ("com.tinkerpop.blueprints","blueprints-core")              => "com.tinkerpop.blueprints" % "blueprints-core" % "2.6.0"
   case ("com.tinkerpop.gremlin","gremlin-groovy")                  => "com.tinkerpop.gremlin" % "gremlin-groovy" % "2.6.0"
   case ("com.thaiopensource","jing")                               => "com.thaiopensource" % "jing" % "20091111"
   case ("com.typesafe","config")                                   => "com.typesafe" % "config" % "1.3.3"
-  case ("com.typesafe.scala-logging","scala-logging")              => "com.typesafe.scala-logging" %% "scala-logging" % "3.9.0"
-  case ("com.typesafe.akka", "akka-stream-kafka")                  => "com.typesafe.akka" %% "akka-stream-kafka" % "0.19"
+  case ("com.typesafe.scala-logging","scala-logging")              => "com.typesafe.scala-logging" %% "scala-logging" % "3.9.2"
+  case ("com.typesafe.akka", "akka-stream-kafka")                  => "com.typesafe.akka" %% "akka-stream-kafka" % "1.0-RC1"
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/build.sbt]: done to 1.0.1

File: server/cmwell-bg/build.sbt [hunk 20]
```diff
@@ -25,14 +25,18 @@ libraryDependencies ++= {
     dm("org.apache.kafka", "kafka")
       .exclude("org.slf4j", "slf4j-log4j12")
       .exclude("log4j", "log4j"),
-    dm("uk.org.lidalia","sysout-over-slf4j")
+    dm("uk.org.lidalia","sysout-over-slf4j"),
+//    dm("io.netty", "netty-transport"),
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-bg/build.sbt]: done

File: server/cmwell-bg/src/main/scala/cmwell/bg/ImpStream.scala [hunk 21]
```diff
@@ -104,23 +104,24 @@ class ImpStream(partition: Int,
...
+            if (createResponse.isAcknowledged) {
                   logger.info(
                     s"successfully created first index for partition $partition"
                   )
-                  scheduleFuture(5.seconds) {
-                    logger.info("updating all aliases")
-                    ftsService.updateAllAlias()
-                  }
+              Future.successful(true)
+//              scheduleFuture(5.seconds) {
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-bg/src/main/scala/cmwell/bg/ImpStream.scala]: done

File: server/cmwell-bg/src/main/scala/cmwell/bg/ImpStream.scala [hunk 22]
```diff
@@ -653,380 +656,377 @@ class ImpStream(partition: Int,
...
+              List(
+                Try {
+                  val indexTime = if (infoton.indexTime.isDefined) None else Some(System.currentTimeMillis())
+                  val infotonWithUpdatedIndexTime =
+                    if (indexTime.isDefined)
+                      infoton.replaceIndexTime(indexTime.get)
+                    else
+                      infoton
+                  val serializedInfoton = JsonSerializerForES.encodeInfoton(infotonWithUpdatedIndexTime, isCurrent)
+                  // TODO: remove after all envs upgraded
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-bg/src/main/scala/cmwell/bg/ImpStream.scala]: done

File: server/cmwell-cons/src/main/scala/ComponentConf.scala [hunk 23]
```diff
@@ -309,38 +305,16 @@ case class ElasticsearchConf(clusterName: String,
...
-        s"-Dcom.sun.management.jmxremote.port=${jmxremotePort}",
-        "-Dcom.sun.management.jmxremote.ssl=false",
-        "-Dcom.sun.management.jmxremote.authenticate=false",
-        "-Delasticsearch",
-        s"-Des.path.home=$home/app/es/cur",
-        s"-Des.config=$home/conf/${dir}/es.yml")
-    }*/
     // Seq(s"-javaagent:$home/app/ctrl/cur", s"-Dctrl.listenAddress=$listenAddress", s"-Dctrl.seedNodes=${host}",
     // s"-Dctrl.clusterName=$clusterName", s"-Dctrl.roles=Metrics,ElasticsearchNode")
     val agentLibArgs = Seq.empty
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-cons/src/main/scala/ComponentConf.scala]: done

File: server/cmwell-cons/src/main/scala/DataInitializer.scala [hunk 24]
```diff
@@ -59,7 +59,7 @@ class DataInitializer(h: Host, jwt: String, rootDigest: String, rootDigest2: Str
     import SimpleResponse.Implicits.UTF8StringHandler
     import scala.concurrent.duration._
 
-    val totalWaitPeriod = 90.seconds
+    val totalWaitPeriod = 300.seconds
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-cons/src/main/scala/DataInitializer.scala]: done

File: server/cmwell-cons/src/main/scala/Deployment.scala [hunk 25]
```diff
@@ -366,16 +366,20 @@ case class ElasticsearchProps(h: Host)
     with RunnableComponent {
   override val componentName: String = "elasticsearch"
   override val componentDataDirs: Map[String, GenSeq[String]] = Map(
-    "es" -> h.getDataDirs.esDataDirs,
-    "es-master" -> List(s"${h.getInstDirs.intallationDir}/data")
+    "es" -> h.getDataDirs.esDataDirs.filterNot(_.endsWith("master")),
+//    "es-master" -> List(s"${h.getInstDirs.intallationDir}/data")
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-cons/src/main/scala/Deployment.scala]: done

File: server/cmwell-cons/src/main/scala/ctl.scala [hunk 26]
```diff
@@ -554,7 +562,7 @@ abstract class Host(user: String,
       // scalastyle:off
       if (verbose) println("command: " + cmd.mkString(" "))
       // scalastyle:on
-      (s"echo -e -n $pass\\n" #| cmd).!!
+      (Seq("bash", "-c", s"echo -e -n $pass\\\\n") #| cmd).!!
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-cons/src/main/scala/ctl.scala]: With pleasure :smile:

File: server/cmwell-cons/src/main/scala/ctl.scala [hunk 27]
```diff
@@ -1598,27 +1621,33 @@ abstract class Host(user: String,
...
+      case Success(res) =>
+        if (res.trim != """{"acknowledged":true,"shards_acknowledged":true,"index":"cm_well_p0_0"}""") {
+          println(s"Elasticsearch first index creation failed. The response was: $res")
+          sys.exit(1)
+        }
+      case Failure(ex) =>
+        println(s"Elasticsearch first index creation failed with: $ex")
+        sys.exit(1)
+    }
+//    command(s"curl -s -X POST http://${pingAddress}:$esRegPort/cm_well_p0_0/", hosts(0), false)
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-cons/src/main/scala/ctl.scala]: done

File: server/cmwell-fts/build.sbt [hunk 28]
```diff
@@ -22,7 +21,9 @@ libraryDependencies ++= {
 		dm("com.google.code.findbugs", "jsr305"),
 		dm("com.typesafe", "config"),
 		dm("com.spatial4j", "spatial4j"),
-		dm("org.slf4j","jcl-over-slf4j"))
+		dm("org.slf4j","jcl-over-slf4j"),
+		dm("org.apache.logging.log4j", "log4j-to-slf4j"),
+    dm("net.lingala.zip4j", "zip4j") % Test)
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/build.sbt]: done

File: server/cmwell-fts/build.sbt [hunk 29]
```diff
@@ -1,18 +1,17 @@
...
 	Seq(
 		dm("org.elasticsearch", "elasticsearch")
-			.exclude("org.slf4j", "slf4j-log4j12")
-			.exclude("log4j", "log4j")
 			.exclude("com.fasterxml.jackson.core", "jackson-core")
 			.exclude("com.google.guava", "guava")
 			.exclude("org.apache.lucene", "lucene-queryparser")
 			.exclude("org.apache.lucene", "lucene-analyzers-common")
 			.exclude("org.apache.lucene", "lucene-core") jar(),
+		dm("org.elasticsearch.client", "transport"),
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/build.sbt]: correct! removed all the other dependencies

File: server/cmwell-fts/src/main/resources/indices_template.json [hunk 30]
```diff
@@ -1,17 +1,22 @@
 {
-  "template": "*cmwell*",
+  "index_patterns": "cm_well_*",
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/src/main/resources/indices_template.json]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 31]
```diff
@@ -655,66 +569,109 @@ class FTSServiceNew(config: Config, esClasspathYaml: String)
...
+
+  def delete(uuid:String)(implicit executionContext:ExecutionContext):Future[Boolean] = {
+    injectFuture[SearchResponse](client.prepareSearch().setQuery(idsQuery("infoclone").addIds(uuid)).execute(_)).collect[String]{
+      case res if res.getHits.getHits.nonEmpty => res.getHits.getHits.head.getIndex
+    }.flatMap{ indexName =>
+      delete(uuid, indexName)
+    }.recover{case _ => false}
+  }
+
+  private def startShardScrollEliNew(pathFilter: Option[PathFilter] = None, fieldsFilter: Option[FieldFilter] = None,
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 32]
```diff
@@ -655,66 +569,109 @@ class FTSServiceNew(config: Config, esClasspathYaml: String)
...
+      val ftsResponse = FTSScrollResponse(scrollResponse.getHits.totalHits,
+        scrollResponse.getScrollId,
+        esResponseToInfotons(scrollResponse, includeScore = false),
+        Some(nodeId))
+      FTSStartScrollResponseEliNew(ftsResponse, searchQueryStr = searchQueryStr)
     }
   }
 
-  def startSuperScroll(
+  def startSuperScrollEliNew(
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 33]
```diff
@@ -655,66 +569,109 @@ class FTSServiceNew(config: Config, esClasspathYaml: String)
...
+
+    val searchQueryStr = if (debugInfo) Some(request.toString) else None
 
     scrollResponseFuture.map { scrollResponse =>
-      FTSStartScrollResponse(scrollResponse.getHits.totalHits, scrollResponse.getScrollId, Some(nodeId))
+      val ftsResponse = FTSScrollResponse(scrollResponse.getHits.totalHits,
+        scrollResponse.getScrollId,
+        esResponseToInfotons(scrollResponse, includeScore = false),
+        Some(nodeId))
+      FTSStartScrollResponseEliNew(ftsResponse, searchQueryStr = searchQueryStr)
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 34]
```diff
@@ -655,66 +569,109 @@ class FTSServiceNew(config: Config, esClasspathYaml: String)
...
+  def startSuperScrollEliNew(
     pathFilter: Option[PathFilter],
     fieldsFilter: Option[FieldFilter],
     datesFilter: Option[DatesFilter],
     paginationParams: PaginationParams,
     scrollTTL: Long,
     withHistory: Boolean,
     withDeleted: Boolean
-  )(implicit executionContext: ExecutionContext): Seq[() => Future[FTSStartScrollResponse]] = {
+  )(implicit executionContext: ExecutionContext): Seq[() => Future[FTSStartScrollResponseEliNew]] = {
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 35]
```diff
@@ -750,81 +707,55 @@ class FTSServiceNew(config: Config, esClasspathYaml: String)
...
-    paginationParams: PaginationParams,
-    scrollTTL: Long,
-    withHistory: Boolean,
-    withDeleted: Boolean,
-    indexNames: Seq[String],
-    onlyNode: Option[String],
-    partition: String,
-    debugInfo: Boolean
-  )(implicit executionContext: ExecutionContext, logger: Logger = loger): Future[FTSStartScrollResponse] = {
+  def startScrollEliNew(pathFilter: Option[PathFilter],
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 36]
```diff
@@ -750,81 +707,55 @@ class FTSServiceNew(config: Config, esClasspathYaml: String)
...
+                  indexNames: Seq[String] = Seq.empty,
+                  onlyNode: Option[String] = None,
+                  partition: String = defaultPartition,
+                  debugInfo: Boolean = false)
+                 (implicit executionContext:ExecutionContext, logger:Logger = loger) : Future[FTSStartScrollResponseEliNew] = {
     logger.debug(s"StartScroll request: $pathFilter, $fieldsFilter, $datesFilter, $paginationParams, $withHistory")
 
     val fields = "system.kind" :: "system.path" :: "system.uuid" :: "system.lastModified" :: "content.length" ::
-      "content.mimeType" :: "link.to" :: "link.kind" :: "system.dc" :: "system.indexTime" :: "system.quad" :: Nil
+      "content.mimeType" :: "link.to" :: "link.kind" :: "system.dc" :: "system.indexTime" :: "system.quad" :: "system.current" :: Nil
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 37]
```diff
@@ -750,81 +707,55 @@ class FTSServiceNew(config: Config, esClasspathYaml: String)
...
-    pathFilter: Option[PathFilter],
-    fieldsFilter: Option[FieldFilter],
-    datesFilter: Option[DatesFilter],
-    paginationParams: PaginationParams,
-    scrollTTL: Long,
-    withHistory: Boolean,
-    withDeleted: Boolean,
-    partition: String
-  )(implicit executionContext: ExecutionContext, logger: Logger = loger): Seq[Future[FTSStartScrollResponse]] = {
+/*
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 38]
```diff
@@ -859,17 +790,16 @@ class FTSServiceNew(config: Config, esClasspathYaml: String)
...
-    pathFilter: Option[PathFilter],
-    fieldsFilter: Option[FieldFilter],
-    datesFilter: Option[DatesFilter],
-    paginationParams: PaginationParams,
-    scrollTTL: Long,
-    withHistory: Boolean,
-    withDeleted: Boolean,
-    partition: String
-  )(implicit executionContext: ExecutionContext, logger: Logger = loger): Seq[Future[FTSStartScrollResponse]] = {
+  def startMultiScrollEliNew(pathFilter: Option[PathFilter], fieldsFilter: Option[FieldFilter],
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 39]
```diff
@@ -1661,101 +1519,96 @@ class FTSServiceNew(config: Config, esClasspathYaml: String)
...
     }
   }
 
-  override def purgeHistory(path: String, isRecursive: Boolean, partition: String)(
-    implicit executionContext: ExecutionContext
-  ): Future[Boolean] = ???
-
-  override def getMappings(withHistory: Boolean,
-                           partition: String)(implicit executionContext: ExecutionContext): Future[Set[String]] = {
+/*
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 40]
```diff
@@ -1777,28 +1630,51 @@ class FTSServiceNew(config: Config, esClasspathYaml: String)
...
     }
-
-    logRequest("purge", s"uuid: $uuid")
-    injectFuture[BulkResponse](bulkRequest.execute(_)).map(_ => true)
   }
 
-  override def purgeByUuidsFromAllIndexes(uuids: Vector[String], partition: String)(
-    implicit executionContext: ExecutionContext
-  ): Future[BulkResponse] = {
+  def countSearchOpenContexts(): Array[(String, Long)] = ???
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: It's part of `search-contexts.md`.

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 41]
```diff
@@ -1777,28 +1630,51 @@ class FTSServiceNew(config: Config, esClasspathYaml: String)
...
+      None,
+      Some(MultiFieldFilter(Must, fieldFilters.fold(filtersSeq)(filtersSeq.::))),
+      None,
+      withHistory = withHistory
+    )
+    logRequest("getLastIndexTimeFor", s"dc: $dc")
+    injectFuture[SearchResponse](request.execute).map(_.getHits.getHits.headOption.map(_.getFields.get("system.indexTime").getValue[Long]))
+  }
+
+//  def purgeByUuidsFromAllIndexes(uuids: Vector[String], partition: String)(implicit executionContext: ExecutionContext): Future[BulkResponse] = ???
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 42]
```diff
@@ -1922,3 +1794,171 @@ case class BulkResult(successful: Iterable[String],
...
+}
+
+case class DatesFilter(from: Option[DateTime], to:Option[DateTime])
+case class PaginationParams(offset: Int, length: Int)
+object DefaultPaginationParams extends PaginationParams(0, 100)
+
+case class FTSSearchResponse(total: Long, offset: Long, length: Long, infotons: Seq[Infoton], searchQueryStr: Option[String] = None)
+case class FTSStartScrollResponseEliNew(response: FTSScrollResponse, searchQueryStr: Option[String] = None)
+case class FTSScrollResponse(total: Long, scrollId: String, infotons: Seq[Infoton], nodeId: Option[String] = None, searchQueryStr: Option[String] = None)
+//case class FTSScrollThinResponse(total: Long, scrollId: String, thinInfotons: Seq[FTSThinInfoton], nodeId: Option[String] = None)
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-irw/build.sbt [hunk 43]
```diff
@@ -6,24 +6,25 @@ libraryDependencies ++= {
   val dm = dependenciesManager.value
   Seq(
     dm("org.slf4j", "slf4j-api"),
-    dm("commons-codec", "commons-codec"),
+//    dm("commons-codec", "commons-codec"),
```
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-irw/build.sbt]: done

File: server/cmwell-ws/app/cmwell/ws/Streams.scala (hunk 2 above)
Comment (e-orz) [reply server/cmwell-ws/app/cmwell/ws/Streams.scala]: done

File: server/cmwell-ws/app/cmwell/ws/Streams.scala (hunk 3 above)
Comment (e-orz) [reply server/cmwell-ws/app/cmwell/ws/Streams.scala]: done

File: server/cmwell-ws/app/ld/cmw/TimeBasedAccumulatedNamespacesCache.scala (hunk 5 above)
Comment (e-orz) [reply server/cmwell-ws/app/ld/cmw/TimeBasedAccumulatedNamespacesCache.scala]: done

File: server/cmwell-ws/app/logic/CRUDServiceFS.scala (hunk 6 above)
Comment (e-orz) [reply server/cmwell-ws/app/logic/CRUDServiceFS.scala]: done

File: server/cmwell-ws/app/logic/CRUDServiceFS.scala (hunk 7 above)
Comment (e-orz) [reply server/cmwell-ws/app/logic/CRUDServiceFS.scala]: done

File: server/cmwell-ws/src/pack/application.conf (hunk 8 above)
Comment (e-orz) [reply server/cmwell-ws/src/pack/application.conf]: done

**Pull request merged**